rfd/0030-upgrade-system.md

```diff
@@ -0,0 +1,444 @@
+may not be practical.  If the latest version turns out to have an unexpected issue, a downgrade may be
+necessary.  The simplest solution (and the one we'll likely start with) is to require that downgrades
+be triggered manually, but this isn't ideal.  A better solution would be to have custom policies for
+downgrades, with a reasonable default being something like "previous patch release iff current release
+was not a security patch".  Complementary to this would be including an ability to yank outdated/insecure
+versions from distribution (or at least mark them as such).
+
+Masked downgrades aren't a significant threat if only the distribution system is compromised, as the
+index/manifest system will continue to provide checksums with correct versioning information.  Some
+repository systems also resist tampering with version numbers, though not all do so.  That leaves us
+with one final scenario: the index *and* at least one repository are compromised simultaneously (but
+the compromise is not so severe as to put us in the previously discussed 'doomsday scenario').  This is
+a bit niche, but lets exercise a little professional paranoia.  We should have a mechanism of validating
+package version (prior to unpacking/installation) as part of package signature verification (i.e. on the
+"distribution" side).  One appealing option is [minisign](https://jedisct1.github.io/minisign/) which
+is a simple and modern signing utility that supports "trusted comments" (i.e. signed metadata) as part
+of the signature file.
+
+
+#### Compromised Build System
+
+Out of scope for now, but I've heard all the cool kids are doing deterministic builds these days.
+
+
+## Proposal
+
+### Overview
+
+High-level overview of system components, including discussion of MVP vs planned features.
+
+#### Index Server
+
+The index server, as discussed above, is responsible for serving metadata and checksums for the
+current list of valid installation targets.  Part of the security model of the upgrade system is
+that the index server remains isolated from the rest of the ugrade system stack, and provides only
+abstract information about available releases (i.e. it will not provide things like download urls
+or install scripts).
+
+Key design considerations related to the index server are as follows:
+- Index server should be trivially horizontally scalable.
+- Index servers should always require TLS 1.3 unless started in development mode.
+- Rollbacks of index server state or API version must not be detremental to its clients
+(i.e. clients must be resilient to inconsistent values being returned).
+- Index server must remain compatible with many past teleport versions (exact number TBD),
+but we don't want to orphan older versions, so lets err on the side of caution here.
+
+Outstanding design questions:
+- Grpc or basic https API?  My preference leans toward grpc, but its hard to discount the possible
```

Comment:
I would standardize on GRPC - Teleport is already migrating to GRPC cluster-wide, so seems like a good fit.